runtime/src/main/java/org/corfudb/runtime/collections/StreamManager.java

```diff
@@ -0,0 +1,219 @@
+import org.corfudb.runtime.exceptions.SubscriberAlreadyExistsException;
+import org.corfudb.runtime.view.Address;
+import org.corfudb.runtime.view.ObjectsView;
+import org.corfudb.runtime.view.StreamOptions;
+import org.corfudb.runtime.view.TableRegistry;
+import org.corfudb.runtime.view.stream.IStreamView;
+
+/**
+ *
+ * Created by hisundar on 04/28/2020.
+ */
+@Slf4j
+public class StreamManager {
+
+    private final int MAX_SUBSCRIBERS = 12; // Arbitrary sanity check value.
+
+    /**
+     * Corfu Runtime.
+     */
+    @Getter
+    private final CorfuRuntime runtime;
+
+    /**
+     * Map of StreamingSubscriptionThreads.
+     * <p>
+     * This is the map of all streaming subscriptions.
+     */
+    @Getter
+    private final Map<StreamListener, SubscriptionThread> subscriptions = new HashMap<>();
+
+    public StreamManager(@Nonnull CorfuRuntime runtime) {
+        this.runtime = runtime;
+    }
+
+    /**
+     * Subscribe to updates.
+     *
+     * @param streamListener   Client listener.
+     * @param namespace        Namespace of interest.
+     * @param tablesOfInterest Only updates from these tables will be returned.
+     * @param startAddress     Address to start the notifications from.
+     */
+    public synchronized <K extends Message, V extends Message, M extends Message>
+    void subscribe(@Nonnull StreamListener streamListener, @Nonnull String namespace,
+                   @Nonnull List<TableSchema> tablesOfInterest, long startAddress) {
+        if (subscriptions.size() >= MAX_SUBSCRIBERS) {
+            throw new SubscriberAlreadyExistsException(
+                    "subscribe: Max "+MAX_SUBSCRIBERS+" hit! Consider consolidating "+streamListener);
+        }
+        if (subscriptions.containsKey(streamListener)) {
+            throw new SubscriberAlreadyExistsException(
+                    "StreamManager:subscrbie already registered "+streamListener+". Maybe call unregister");
+        }
+        log.info("StreamManager::subscribe {}, startAddress {}, namespace {}, tables {}", streamListener,
+                startAddress, namespace, tablesOfInterest.toString());
+        SubscriptionThread st = new SubscriptionThread(this, streamListener,
+                namespace, tablesOfInterest, startAddress);
+        subscriptions.put(streamListener, st);
+        st.start();
+    }
+
+    /**
+     * Unsubscribe a prior subscription.
+     *
+     * @param streamListener Client listener.
+     */
+    public synchronized void unsubscribe(@Nonnull StreamListener streamListener) {
+        SubscriptionThread st = subscriptions.remove(streamListener);
+        if (st == null) {
+            log.warn("StreamManager::unsubscribe has no context for "+streamListener.toString());
+            return;
+        }
+        st.setShutdown(true);
+        st.interrupt();
+        log.info("Unsubscribed StreamListener {}", streamListener.toString());
+    }
+
+    class SubscriptionThread extends Thread {
```

Comment:
I'm concerned, probably we gonna create too many threads

Reply:
We have a cap on it at MAX_SUBSCRIBERS currently set to an arbitrary value of 12.

Reply:
Also we are talking about 1 thread per business vertical so number of threads is currently just one. However one can argue that even if there are multiple threads since they correspond to actual business verticals they should be fully justified.